### ISSUE TITLE
[embedder] Need a mechanism to configure microtasks to run on the platform thread.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#152337]
The iOS/Android embedders handle this by setting up fml::MessageLoop on the platform thread, but there is no equivalent API for embedders, which prevents us from making ui thread == platform thread on desktop.

See: https://github.com/flutter/engine/pull/53747

@chinmaygarde

**Comment by knopp** [comment]
Does this even need to be an API? Can the engine just flush the microtasks at the end of scheduled block? Seems to me that this is what is happening anyway, just done in a somewhat convoluted way?

**Comment by chinmaygarde** [comment]
@zanderso Did you mean to assign this to me? Per our chat, I'm still heads down on the QNX toolchain patches. Will take a look after though.

**Comment by zanderso** [comment]
> Did you mean to assign this to me? Per our chat, I'm still heads down on the QNX toolchain patches. Will take a look after though.

Yeah sorry if I'm jumping the gun a bit. The ordering we talked about earlier today sgtm.

**Comment by knopp** [comment]
This has a pretty straightforward solution, adding `FlutterEngineFlushMicrotaskQueue` to embedder API, which flushes the root isolate microtask queue. WIP PR that uses this: https://github.com/flutter/flutter/pull/162671

**Comment by chinmaygarde** [comment]
@knopp I haven't paged this in yet so I may be missing something. But my expectation was that no embedder API changes for microtasks will be necessary if we allowed embedders to specify a UI task runner in addition to a render and platform task runners. Then the UI thread task that is now embedder responsibility to invoke (via `FlutterEngineRunTask`) will also flush microtasks. Microtask execution should be entirely seamless for the embedders. They only care about providing  an additional UI task runner (which happens to run on the main thread).

Making the embedder do something special for microtasks just because the UI thread is the same as the platform thread looks confusing to the embedder implementer.

**Comment by knopp** [comment]
@chinmaygarde, agreed. `FlutterEngineFlushMicrotaskQueue` is good enough for the prototype to work, but in the end it should be responsibility of `EmbedderEngine` and not the embedder implementation. I think `EmbedderEngine::RunTask` might be good place to determine whether the microtask task queue needs to be flushed and to flush it if it does.
